### PR TITLE
fixes #92: Provide Kafka Connect Sink

### LINKED ISSUE #92: Provide Kafka Connect Sink
## Feature description (Mandatory)
Add a Kafka Connect Sink Plugin, that can be used inside the Confluent Platform

## How this feature can improve the project?
This let Kafka users to simply integrate Neo4j into their streaming flow

### PR DESCRIPTION
**Pull request by conker84**
Fixes #92 

First iteration of the Kafka Connect Sink

## Proposed Changes (Mandatory)

A brief list of proposed changes in order to fix the issue:

  - Created a package common in order to share functionalities between modules
  - Create a subproject `kafka-connect-neo4j` with the current implementation
  - Added docker compose file in order to test it

### CODE REVIEW

File: common/src/main/kotlin/streams/utils/JacksonUtils.kt [hunk 1]
```diff
@@ -68,4 +68,9 @@ object JacksonUtil {
         return OBJECT_MAPPER
     }
 
+    fun asMap(any: Any): Map<String, Any?> {
+        return ObjectMapper().convertValue(any, Map::class.java)
```
[the remark this answers is not in the document]
Comment (conker84) [reply common/src/main/kotlin/streams/utils/JacksonUtils.kt]: My bad sorry

File: common/src/test/kotlin/streams/utils/JacksonUtilsTest.kt [hunk 2]
```diff
@@ -27,7 +27,7 @@ class JacksonUtilTest {
                 "dateTime" to datetime(date( 2017, 12, 17), time( 17, 14, 35, 123456789, UTC)))
 
         // When
-        val jsonString = JacksonUtil.getMapper().writeValueAsString(map)
+        val jsonString = JacksonUtils.getMapper().writeValueAsString(map)
```
[the remark this answers is not in the document]
Comment (conker84) [reply common/src/test/kotlin/streams/utils/JacksonUtilsTest.kt]: Ok

File: kafka-connect-neo4j/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# Introduction
+
+Welcome to your new Kafka Connect connector!
+
+# Running in development
+
+
+The [docker-compose.yml](docker-compose.yml) that is included in this repository is based on the Confluent Platform Docker
```
[the remark this answers is not in the document]
Comment (conker84) [reply kafka-connect-neo4j/README.md]: Ok

File: kafka-connect-neo4j/docker/readme.adoc [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    $ mvn clean install
+
+Inside the directory <neo4j-streams>/kafka-connect-neo4j/target` youl'll find a file named `kafka-connect-neo4j-<VERSION>.tar.gz`
+
+
+# Configuring the stack
+
+
+Create a directory `plugin` at the same level of the compose file and unzip the file `kafka-connect-neo4j-<VERSION>.tar.gz` inside it, then start the compose file
```
[the remark this answers is not in the document]
Comment (conker84) [reply kafka-connect-neo4j/docker/readme.adoc]: Ok

File: kafka-connect-neo4j/docker/readme.adoc [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    $ docker-compose up -d
+
+Create the Sink instance:
+
+We'll define the Sink configuration as follows:
+
+```json
```
[the remark this answers is not in the document]
Comment (conker84) [reply kafka-connect-neo4j/docker/readme.adoc]: Ok

File: kafka-connect-neo4j/docker/readme.adoc [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+In particular this line:
+
+```
+"neo4j.topic.cypher.my-topic": "MERGE (p:Person{name: event.name, surname: event.surname}) MERGE (f:Family{name: event.surname}) MERGE (p)-[:BELONGS_TO]->(f)"
+```
+
+defines that all the data that comes from the topic `neo4j` will be unpacked by the Sink into Neo4j with the following Cypher query:
+
+```cypher
+MERGE (p:Person{name: event.name, surname: event.surname})
```
[the remark this answers is not in the document]
Comment (conker84) [reply kafka-connect-neo4j/docker/readme.adoc]: Ok

File: kafka-connect-neo4j/pom.xml [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+    <repositories>
+        <repository>
+            <releases>
+                <enabled>true</enabled>
+            </releases>
+            <snapshots>
+                <enabled>false</enabled>
+            </snapshots>
+            <id>bintray.kotlin.eap</id>
```
[the remark this answers is not in the document]
Comment (conker84) [reply kafka-connect-neo4j/pom.xml]: Ok

File: kafka-connect-neo4j/docker/readme.adoc [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+|Name
+|Topics
+
+|Running
+|1
+|Neo
+|my-topic
+|===
+
+Now you can send your AVRO data to the `my-topic` topic
```
[the remark this answers is not in the document]
Comment (conker84) [reply kafka-connect-neo4j/docker/readme.adoc]: Yes, it's delegated to the Confluent platform the user must provide the appropriate converter.

File: kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/Neo4jSinkTask.kt [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+                                it.runAsync(query, mapOf("events" to events))
+                            }
+                            .whenCompleteAsync { cursor, error ->
+                                countDownLatch.countDown()
+                                if (error != null) {
+                                    log.error("""Error while executing the query: $query""", error)
+                                }
+                            }
+                }
+        countDownLatch.await(this.config.batchTimeout, TimeUnit.SECONDS)
```
[the remark this answers is not in the document]
Comment (conker84) [reply kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/Neo4jSinkTask.kt]: I implemented this pattern, what do you think about it?
https://github.com/neo4j-contrib/neo4j-streams/blob/c8e753f96df2b488f4c4ff436a9340ad76d7a4c9/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/Neo4jService.kt#L82-L99

**Comment by conker84** [comment]
Hi @jexp I updated the PR with your feedbacks!

File: kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/Neo4jSinkConnectorConfig.kt [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+        connectionPoolMaxSize = getInt(CONNECTION_POOL_MAX_SIZE)
+        connectionAcquisitionTimeout = getLong(CONNECTION_MAX_CONNECTION_ACQUISITION_TIMEOUT_MSECS)
+
+        loadBalancingStrategy = ConfigUtils
+                .getEnum(Config.LoadBalancingStrategy::class.java, this, CONNECTION_LOAD_BALANCE_STRATEGY) as Config.LoadBalancingStrategy
+        batchTimeout = getLong(BATCH_TIMEOUT_MSEC)
+
+        topicMap = originals
+                .filterKeys { it.toString().startsWith(TOPIC_CYPHER_PREFIX) }
+                .map { it.key.toString().replace(TOPIC_CYPHER_PREFIX, String.emptyString()) to it.value.toString() }
```
[the remark this answers is not in the document]
Comment (jexp) [reply kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/Neo4jSinkConnectorConfig.kt]: Ok, then it's ok.

File: kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/Neo4jSinkTask.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (jexp) [reply kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/Neo4jSinkTask.kt]: Looks good to me. 
We could ask in the Kotlin slack channel if anyone thinks there is another/better way.

btw teh extracted debug() function only works when it's inline, so perhaps we can force-inline it in kotlin with the `inline` keyword.

File: kafka-connect-neo4j/README.md [hunk 11]
```diff
@@ -0,0 +1,15 @@
+# Introduction
+
+Welcome to your Neo4j Kafka Connect connector!
```
Comment (jexp) [inline kafka-connect-neo4j/README.md]: We need more detail here. But can do this after this PR.

**Pull request merged**